src/Client.php

```diff
@@ -3,63 +3,270 @@
     {
         // TODO: Implement changePassword() method.
     }
 
+    /**
+     * @inheritDoc
+     */
     public function changeAttribute(string $identifier, string $key, string $value): bool
     {
         // TODO: Implement changeAttribute() method.
     }
 
+    /**
+     * @inheritDoc
+     */
     public function changeAttributes(string $identifier, array $attributes): bool
     {
         // TODO: Implement changeAttributes() method.
     }
 
+    /**
+     * @inheritDoc
+     */
     public function enableUser(string $identifier): bool
     {
         // TODO: Implement enableUser() method.
     }
 
+    /**
+     * @inheritDoc
+     */
     public function disableUser(string $identifier): bool
     {
         // TODO: Implement disableUser() method.
     }
 
+    /**
+     * @inheritDoc
+     */
     public function getResourceOwner(AccessTokenInterface $token): ResourceOwnerInterface
     {
-        // TODO: Implement getResourceOwner() method.
+        // If the ID token is not null, use to build the resource owner.
+        // Otherwise, use the claims from the access token.
+        if ($idToken = $token->getIdToken()) {
+            $tokenClaims = $this->decodeToken($idToken);
+        } else {
+            $tokenClaims = $this->decodeToken($token->getToken());
+        }
+
+        return $this->createResourceOwner($tokenClaims, $token);
+    }
+
+    protected function createResourceOwner(array $claims, AccessTokenInterface $token): ResourceOwnerInterface
+    {
+        return new LdapUser($claims);
     }
 
+    /**
+     * @inheritDoc
+     */
     public function decodeToken(string $token): array
     {
-        // TODO: Implement decodeToken() method.
+        return $this->getTokenFactory()->validate($token);
     }
 
+    /**
+     * @inheritDoc
+     */
     public function refreshTokens(string $refreshToken, string $identifier): AccessTokenInterface
     {
-        // TODO: Implement refreshTokens() method.
+        throw new ClientException('Refreshing of tokens is not supported.');
     }
 
+    /**
+     * @inheritDoc
+     */
     public function getUserByIdentifier(string $identifier): ResourceOwnerInterface
     {
-        // TODO: Implement getUserByIdentifier() method.
+        $dn = $this->buildDn($identifier);
+
+        $query = $this->ldap->query($dn, '(objectClass=inetOrgPerson)');
+        $entry = $query->execute();
+
+        if (empty($entry)) {
+            throw new ClientException('User not found.');
+        }
+
+        $user = $entry[0];
+
+        $attributes = [
+            'dn' => $user->getDn(),
+            'cn' => $user->getAttribute('cn'),
+        ];
+
+        return new LdapUser($attributes);
+    }
+
+    public function setTokenFactory(TokenFactoryInterface $factory): self
+    {
+        $this->tokenFactory = $factory;
+
+        return $this;
+    }
+
+    public function getTokenFactory(): TokenFactoryInterface
+    {
+        if (isset($this->tokenFactory)) {
```

Comment:
Any reason we aren't passing this in on the constructor?

Reply:
No reason, either way would be fine, I'm just a fan of doing it like this for truly option parameters. Makes it easier to change during runtime too (generally, but also unit testing).